### PR TITLE
add Pair to filter and filter!

### PR DESCRIPTION
**Pull request by bkamins**
Related to #2088. Easy filters will be faster and easier to write with it.

### CODE REVIEW

**Comment by bkamins** [comment]
@nalimilan - this should be good to have a look at.

As discussed on Slack we have a decision if we want the function to be applied for each row (which is natural here). I think that in the end after all these `Pair` related PRs are merged I will add a section to the manual explaining the `Pair` usage scenarios.

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -863,9 +863,15 @@ end
 
 """
     filter(function, df::AbstractDataFrame)
+    filter(cols => function, df::AbstractDataFrame)
 
 Return a copy of data frame `df` containing only rows for which `function`
-returns `true`. The function is passed a `DataFrameRow` as its only argument.
+returns `true`. The function is passed a `DataFrameRow` as its only argument
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: I'd turn this into three sentences (no cols, single col, multiple cols), each starting with ``If `cols` ...``

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -886,15 +892,40 @@ julia> filter(row -> row[:x] > 1, df)
 ├─────┼───────┼────────┤
 │ 1   │ 3     │ b      │
 │ 2   │ 2     │ a      │
+
+julia> filter(:x => >(1), df)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: As noted in another PR, I'd rather use something which doesn't involve the fancy syntax:
```suggestion
julia> filter(:x => x -> x > 1, df)
```

Also show an example with several columns?

Same below.

File: src/abstractdataframe/abstractdataframe.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: fixed

**Comment by bkamins** [comment]
all comments should be fixed now.

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -866,12 +866,16 @@ end
...
 Return a copy of data frame `df` containing only rows for which `function`
-returns `true`. The function is passed a `DataFrameRow` as its only argument
-if `cols` is not specified.
-If `cols` is passed then elements of column `cols` are passed to the function,
-if `cols` is a single column or as a `NamedTuple` if multiple columns are passed
-(any valid column selector is accepted). Passing `cols` leads to a more efficient
-execution of the operation for large data frames.
+returns `true`.
+
+If `cols` is not specified then he function is passed `DataFrameRow`s as its only argument.
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Mixing plural and singular is weird:
```suggestion
If `cols` is not specified then the function is passed `DataFrameRow`s.
```

**Comment by bkamins** [comment]
Thank you. I have applied the same fix to `filter!` docstring.

**Comment by bkamins** [comment]
This should be good for a final reveiw. Thank you.

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -867,9 +867,19 @@ end
...
     filter(function, df::AbstractDataFrame)
+    filter(cols => function, df::AbstractDataFrame)
 
 Return a copy of data frame `df` containing only rows for which `function`
-returns `true`. The function is passed a `DataFrameRow` as its only argument.
+returns `true`.
+
+If `cols` is not specified then the function is passed `DataFrameRow`s.
+If `cols` is a single column then then elements of column `cols` are passed to the function.
+If `cols` represents multiple columns (any valid column selector is accepted) then
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: This should be kept consistent with `by`/`combine`/`select` (https://github.com/JuliaData/DataFrames.jl/issues/1935): if we decide to splat, we should also do it here. But I guess we can merge the PR as-is, and change the code before the next breaking release if we decide to change other functions too.

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: Agreed. We can wait a few days with merging this I think (when we get the response on #data).

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: I will rewrite it. We will auto-splat

**Comment by bkamins** [comment]
I have implemented auto-splatting for `filter` and `filter!` so that we can have a look how it works.

Here are some benchmarks:
```
julia> df = DataFrame(x = rand(1:10, 10^6), y = rand(1:10, 10^6));

julia> @benchmark filter(row -> row.x == row.y, df)
BenchmarkTools.Trial:
  memory estimate:  94.79 MiB
  allocs estimate:  4999012
  --------------
  minimum time:     140.924 ms (2.87% GC)
  median time:      147.238 ms (2.94% GC)
  mean time:        158.191 ms (2.99% GC)
  maximum time:     230.591 ms (1.84% GC)
  --------------
  samples:          32
  evals/sample:     1

julia> @benchmark filter([:x, :y] => ==, df)
BenchmarkTools.Trial:
  memory estimate:  3.25 MiB
  allocs estimate:  41
  --------------
  minimum time:     3.825 ms (0.00% GC)
  median time:      4.902 ms (0.00% GC)
  mean time:        5.356 ms (6.40% GC)
  maximum time:     11.774 ms (36.95% GC)
  --------------
  samples:          933
  evals/sample:     1

julia> @benchmark df[df.x .== df.y, :]
BenchmarkTools.Trial:
  memory estimate:  2.42 MiB
  allocs estimate:  32
  --------------
  minimum time:     2.098 ms (0.00% GC)
  median time:      2.957 ms (0.00% GC)
  mean time:        3.203 ms (5.94% GC)
  maximum time:     9.557 ms (50.90% GC)
  --------------
  samples:          1553
  evals/sample:     1
```

so we are way faster than old `filter` and only somewhat slower than direct indexing (part of the reason why we are slower is that we have to do the `::Bool` check when applying the function).

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -873,9 +873,9 @@ Return a copy of data frame `df` containing only rows for which `function`
 returns `true`.
 
 If `cols` is not specified then the function is passed `DataFrameRow`s.
-If `cols` is a single column then then elements of column `cols` are passed to the function.
-If `cols` represents multiple columns (any valid column selector is accepted) then
-`NamedTuple`s formed with elements of the selected columns are passed to the function.
+If `cols` argument is passed then it should be valid column selector (additionally
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
If `cols` is specified then it should be a valid column selector
```
(for consistency with the previous sentence)

File: src/abstractdataframe/abstractdataframe.jl [hunk 6]
```diff
@@ -873,9 +873,9 @@ Return a copy of data frame `df` containing only rows for which `function`
...
+allowing column duplicates are allowed if a vector of `Int` or `Symbol is passed),
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
(column duplicates are allowed if a vector of `Int` or `Symbol` is passed),
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -873,9 +873,9 @@ Return a copy of data frame `df` containing only rows for which `function`
...
+the the function is passed elements of the selected columns as consecutive positional arguments.
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
the function is passed elements of the selected columns as separate positional arguments.
```
(I'm not sure "consecutive" adds anything here, we can't pass non-consecutive arguments. :-)

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -909,38 +909,37 @@ julia> filter(:x => x -> x > 1, df)
...
-end
-
-_filter_helper(f, itr) = collect(f(r)::Bool for r in itr)
+Base.filter(f, df::AbstractDataFrame) = _filter_helper(df, f, eachrow(df))
+Base.filter((col, f)::Pair{<:ColumnIndex}, df::AbstractDataFrame) =
+    _filter_helper(df, f, df[!, col])
+Base.filter((cols, f)::Pair{<:AbstractVector{Int}}, df::AbstractDataFrame) =
+    (cdf = _columns(df); _filter_helper(df, f, (cdf[i] for i in cols)...))
+Base.filter((cols, f)::Pair{<:AbstractVector{Symbol}}, df::AbstractDataFrame) =
+  filter([index(df)[col] for col in cols] => f, df)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
    filter([index(df)[col] for col in cols] => f, df)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 9]
```diff
@@ -909,38 +909,37 @@ julia> filter(:x => x -> x > 1, df)
...
+Base.filter((cols, f)::Pair, df::AbstractDataFrame) =
+  filter(index(df)[cols] => f, df)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
    filter(index(df)[cols] => f, df)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 10]
```diff
@@ -909,38 +909,37 @@ julia> filter(:x => x -> x > 1, df)
...
 """
     filter!(function, df::AbstractDataFrame)
 
 Remove rows from data frame `df` for which `function` returns `false`.
 
 If `cols` is not specified then the function is passed `DataFrameRow`s.
-If `cols` is a single column then then elements of column `cols` are passed to the function.
-If `cols` represents multiple columns (any valid column selector is accepted) then
-`NamedTuple`s formed with elements of the selected columns are passed to the function.
+If `cols` argument is passed then it should be valid column selector (additionally
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Same suggestions.

**Comment by nalimilan** [comment]
> (part of the reason why we are slower is that we have to do the `::Bool` check when applying the function).

Shouldn't the compiler be able to optimize this check away?

**Comment by bkamins** [comment]
> Shouldn't the compiler be able to optimize this check away?

It should in theory but it seems this check adds an overhead (and we have to keep it). Anyway - at least there is hope that this will "just improve" in the future versions of Julia.

File: src/abstractdataframe/abstractdataframe.jl (hunk 10 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: All suggestions applied. Thank you!

**Comment by bkamins** [comment]
Before we merge. If I use broadcasting here like:
```
_filter_helper(df, f, cols...) = df[((x...) -> f(x...)::Bool).(cols...), :]
```
then we are much faster and the `Bool` check seems to be properly optimized out (again the only downside is additional compilation cost as broadcasting is a bit more expensive to compile, but I think it should be OK). So do we change it here and in `select` to broadcasting for `ByRow` operations?

**Comment by nalimilan** [comment]
Let's take the same approach as for `ByRow` and have the discussion there.

**Pull request merged**